### PR TITLE
[Agent] Support kafka framework

### PR DESCRIPTION
**Pull request by ascrutae**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#233

### CODE REVIEW

**Comment by ascrutae** [comment]
Here is the report: https://github.com/SkywalkingTest/agent-integration-test-report/blob/master/2018/1/ascrutae/testReport-master-2018-01-23-11-35.md

**Comment by wu-sheng** [comment]
@ascrutae Glad you have the supported versions and test cases. But I still need your Segment sample data description, to make sure the plugin is working in the right way.

**Comment by wu-sheng** [comment]
@ascrutae Do you merge the test cases before plugin merged. There are all 4 kafka cases fail.

**Comment by ascrutae** [comment]
## Producer Span
1. Operation name :  **Kafka/{TopicName}/Producer** , eg. Kafka/test/Producer
2. tags:
```
   mq.broker: '127.0.0.1:9092'
   mq.topic: test
```
3. Spanlayer: MQ
4. ComponentName: KAFKA

## Consumer Span
1. Operation name :  **Kafka/{TopicName}/Consumer** , eg. Kafka/test/Consumer
2. tags:
```
   mq.broker: '127.0.0.1:9092'
   mq.topic: test
```
3. Spanlayer: MQ
4. ComponentName: KAFKA

Here is my expected data: https://github.com/SkywalkingTest/skywalking-agent-testcases/blob/master/kafka-scenario/config/expectedData.yaml

**Comment by wu-sheng** [comment]
@ascrutae Please send a separated pr to update `updated list` and add some FAQ docs about how to make sure Kafka plugin works as expected.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaConsumerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        ConsumerEnhanceRequiredInfo requiredInfo = (ConsumerEnhanceRequiredInfo)objInst.getSkyWalkingDynamicField();
+        requiredInfo.setStartTime(System.currentTimeMillis());
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        Object ret) throws Throwable {
+        Map<TopicPartition, List<ConsumerRecord<?, ?>>> records = (Map<TopicPartition, List<ConsumerRecord<?, ?>>>)ret;
+        //
+        // The entry span will create when the consumer fetch anyone message from kafka cluster, or the span will not create.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaConsumerInterceptor.java]: Change comments to:
> The entry span will only be created when the consumer received at least one message.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaProducerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        MethodInterceptResult result) throws Throwable {
+
+        ContextCarrier contextCarrier = new ContextCarrier();
+
+        ProducerRecord record = (ProducerRecord)allArguments[0];
+        String topicName = (String)((EnhancedInstance)record).getSkyWalkingDynamicField();
+
+        AbstractSpan activeSpan = ContextManager.createExitSpan(OPERATE_NAME_PREFIX + topicName + PRODUCER_OPERATE_NAME_SUFFIX, contextCarrier, (String)objInst.getSkyWalkingDynamicField());
+
+        //set tags
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaProducerInterceptor.java]: Please remove these meaningless comments.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaProducerInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        Tags.MQ_BROKER.set(activeSpan, (String)objInst.getSkyWalkingDynamicField());
+        Tags.MQ_TOPIC.set(activeSpan, topicName);
+        SpanLayer.asMQ(activeSpan);
+        activeSpan.setComponent(ComponentsDefine.KAFKA);
+
+        // set headers
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/KafkaProducerInterceptor.java]: Same story, no need.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/ProducerConstructorInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+import org.apache.skywalking.apm.util.StringUtil;
+
+public class ProducerConstructorInterceptor implements InstanceConstructorInterceptor {
+
+    @Override public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        ProducerConfig config = (ProducerConfig)allArguments[0];
+        // set the bootstrap server address
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/ProducerConstructorInterceptor.java]: The comments are exact same as codes.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/ProducerConstructorInterceptor.java]: No need either.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/ProducerRecordConstructorInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.apache.skywalking.apm.plugin.kafka.v11;
+
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+
+public class ProducerRecordConstructorInterceptor implements InstanceConstructorInterceptor {
+    @Override public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        String topic = (String)allArguments[0];
+        // set the topic
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/ProducerRecordConstructorInterceptor.java]: `Topic` already is the variable name.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * {@link KafkaProducerInstrumentation} intercept the method <code>send</code> in the class
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: *Instrument didn't intercept anything.They just define interceptors.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/CallbackInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.HierarchyMatch.byHierarchyMatch;
+
+/**
+ * {@link CallbackInstrumentation} intercept the method onCompletion in the class <code>org.apache.kafka.clients.producer.Callback</code>.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/CallbackInstrumentation.java]: *Instrument didn't intercept anything.They just define interceptors.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+ * <code>org.apache.kafka.clients.producer.KafkaProducer</code>. Here is the intercept process steps.
+ *
+ *
+ * <pre>
+ *  1. Record the topic when the client call <code>subscribed</code>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: Is `subscribed` a method? If so, should `xxxx invoked`.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ *  3. Create the entry span when the client call the method <code>pollOnce</code>.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: Same as ^^^

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java (hunk 8 above)
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: And where is No **2**?

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+ *  4. Inject all the <code>Trace Context</code> by iterate all <code>ConsumerRecord</code>
+ *  5. Stop the entry span when end the <code>pollOnce</code> method.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: when `pollOnce` method finish.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * {@link KafkaProducerInstrumentation} intercept the method <code>send</code> in the class
+ * <code>org.apache.kafka.clients.producer.KafkaProducer</code>. Here is the intercept process steps.
+ *
+ *
+ * <pre>
+ *  1. Record the topic when the client call <code>subscribed</code>
+ *  3. Create the entry span when the client call the method <code>pollOnce</code>.
+ *  4. Inject all the <code>Trace Context</code> by iterate all <code>ConsumerRecord</code>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: At consumer side, I think should extract context from the messages.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+ */
+public class KafkaConsumerInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {
+
+    public static final String CONSTRUCTOR_INTERCEPT_FLAG = "org.apache.kafka.clients.consumer.ConsumerConfig";
+    public static final String CONSTRUCTOR_INTERCEPTOR_CLASS = "org.apache.skywalking.apm.plugin.kafka.v11.ConsumerConstructorInterceptor";
+    public static final String INTERCEPTOR_CLASS = "org.apache.skywalking.apm.plugin.kafka.v11.KafkaConsumerInterceptor";
+    public static final String ENHANCE_METHOD = "pollOnce";
+    public static final String ENHANCE_CLASS = "org.apache.kafka.clients.consumer.KafkaConsumer";
+    public static final String SUBSCRIBE_METHOD = "subscribe";
+    public static final String SUBSCRIBE_INTERCEPT_FLAG = "org.apache.kafka.clients.consumer.ConsumerRebalanceListener";
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaConsumerInstrumentation.java]: Why a listener is a flag?

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * {@link KafkaProducerInstrumentation} intercept the method <code>send</code> in the class
+ * <code>org.apache.kafka.clients.producer.KafkaProducer</code>. Here is the intercept process steps.
+ *
+ *
+ * <pre>
+ *  1. Record the broker address when the client create the <code>org.apache.kafka.clients.producer.KafkaProducer</code>
+ * instance
+ *  2. Fetch the topic name from <code>org.apache.kafka.clients.producer.ProducerRecord</code> when the client call
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java]: Same, don't say, client call method. The method is inside the client. Method should be invoked.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+ * <code>send</code> method.
+ *  3. Create the exit span when the client call <code>send</code> method
+ *  4. Set the <code>Context</code> into the <code>org.apache.kafka.clients.producer.ProducerRecord#headers</code>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java]: At here, you should say `Inject`. Please follow our core concepts.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+ *  5. Stop the exit span when end the <code>send</code> method.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/KafkaProducerInstrumentation.java]: Change the sentence too.

File: apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/ProducerRecordInstrumentation.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * {@link ProducerRecordInstrumentation} intercept the constructor in the class <code>org.apache.kafka.clients.producer.ProducerRecord</code>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/kafka-0.11.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/kafka/v11/define/ProducerRecordInstrumentation.java]: *Instrument didn't intercept anything.They just define interceptors.

**Comment by ascrutae** [comment]
Done!

**Pull request merged**